feat: Module Loading Improvements

kelindi opened the pull request:
For new clones:

- We only acquire the lock when we need to clone a new repository
- Added a double-check after acquiring the lock to see if another process created the repo while we were waiting
- Properly release the lock in all error cases
- The lock is held only during the clone/checkout process

For existing repositories:

- Don't acquire a lock
- We simply verify that the HEAD is at the correct commit hash
- If the repository exists but is corrupt or has the wrong hash, we remove it and retry
- This makes the common case (using an already cloned repo) very fast and lock-free

Testing:
```bash
go test -v -tags=unit ./pkg/module -run TestTryLock
```

kelindi committed:
anura rate limiter

kelindi committed:
removed exempt ip references

kelindi committed:
removed unauthorized ratelimiting

kelindi committed:
removed concurrency

kelindi committed:
updated tests

kelindi committed:
updated tests

kelindi committed:
saving by branch + retries

kelindi committed:
merge

kelindi committed:
duplicate fix

hollygrimm committed:
feat: Implement repository locking mechanism with tests

- Introduced tryLock and tryLockWithTimeout functions to manage repository locking.
- Added unit tests for lock acquisition, contention, and stale lock scenarios in utils_test.go.
- Enhanced CloneModule function to utilize the locking mechanism during repository cloning.

Signed-off-by: Holly Grimm <holly.grimm@gmail.com>

hollygrimm committed:
refactor: Improve logging and error handling in module utilities

hollygrimm committed:
feat: Add minimum version check (#544)

* feat: Add supported version helper and version config

* feat: Test version check and config

* feat: Add SERVER_MINIMUM_VERSION config

* feat: Add version config to solver server

* feat: Add version check on resource offer and job offer posts

hollygrimm committed:
Merge remote-tracking branch 'origin/main' into kelindi/module-loading

hollygrimm committed:
feat: Enhance repository locking and error handling

- Improved the locking mechanism in tryLockWithTimeout to include hostname and timestamp in the lock file for better debugging.
- Added tests for various locking scenarios, including concurrent locking attempts and handling stale locks.
- Enhanced error handling and logging in CloneModule to provide clearer feedback during repository operations.
- Implemented retries for cloning and fetching references to improve robustness.

hollygrimm committed:
refactor: Remove unused getRepoLocalPath function from utils

hollygrimm committed:
refactor: Change log level from Info to Debug in CloneModule and resolveToCommitHash functions

bgins commented on pkg/module/utils.go:
```diff
@@ -68,72 +90,412 @@ func ProcessModule(module data.ModuleConfig) (data.ModuleConfig, error) {
+	maxRetries := 3
+	var lastErr error
+
+	for retry := 0; retry < maxRetries; retry++ {
```
Could we convert this to a `range` loop over `maxRetries`?

bgins commented on pkg/module/utils.go:
```diff
@@ -68,72 +90,412 @@ func ProcessModule(module data.ModuleConfig) (data.ModuleConfig, error) {
+				}
+
+				// Check if the lock is too old (> 10 minutes)
+				lockTime, parseErr := time.Parse(time.RFC3339, timestamp)
```
Should we handle the case where a `parseErr` is returned?

bgins commented on pkg/module/utils.go:
```diff
@@ -249,11 +622,131 @@ func LoadModule(module data.ModuleConfig, inputs map[string]string) (*data.Modul
+	var repo *git.Repository
+	var cloneErr error
+
+	for retry := 0; retry < 3; retry++ {
```
Same as above. Could we modernize to a range loop?

hollygrimm committed:
Merge remote-tracking branch 'origin/main' into kelindi/module-loading

hollygrimm committed:
refactor: Simplify retry loop syntax

hollygrimm committed:
test: Remove lockFile on malformed timestamp; Add unit test for handling invalid timestamp in lock file

bgins commented on pkg/module/utils_test.go:
```diff
@@ -57,3 +61,154 @@ func TestSubst(t *testing.T) {
+		lockFile := filepath.Join(tempDir, ".lilypad.lock")
+
+		// Create an old lock file with valid PID but old timestamp
+		oldTime := time.Now().Add(-11 * time.Minute).Format(time.RFC3339)
```
Ha great trick adding -11, love it. ❤️

bgins commented on pkg/module/utils_test.go:
```diff
@@ -57,3 +61,154 @@ func TestSubst(t *testing.T) {
+		var wg sync.WaitGroup
+
+		// Launch 5 goroutines to try to get the lock
+		for i := 0; i < 5; i++ {
```
Could we convert this one for a for range loop as well?

bgins commented on pkg/module/utils_test.go:
```diff
@@ -57,3 +61,154 @@ func TestSubst(t *testing.T) {
+		assert.NoError(t, err, "Should create concurrent test directory")
+
+		// This will track which goroutine got the lock
+		gotLock := make(chan int, 5)
```
Might be good to put this `5` in a variable if we expect it to be the same used in the loop a couple lines below.

hollygrimm committed:
feat: Introduce configurable timeout options for module operations

- Added ModuleTimeoutConfig struct to manage timeout settings for locking and git operations.
- Implemented getDefaultOptionInt helper function to retrieve integer values from environment variables.
- Updated tryLockWithTimeout and CloneModule functions to utilize configurable timeouts.
- Added ConfigureTimeouts and ResetTimeouts functions for dynamic timeout management.
- Renamed TestTryLock to TestTryLockWithTimeout for clarity in unit tests.

hollygrimm committed:
test: Update TestTryLockWithTimeout to use configurable goroutine count

bgins reviewed (approved):
Looks good! Let's send it 🎉

hollygrimm committed:
Merge remote-tracking branch 'origin/main' into kelindi/module-loading
